File: src/data/courses/dsa/dynamic-programming/state-decision-dp/maximum-product-subarray/code/python/maximum_product_subarray.py
class Solution:
    def maxProduct(self,nums)->int:
        max_p=cur_max=cur_min=nums[0]
        for i in range(1,len(nums)):
            if nums[i]<0:cur_max,cur_min=cur_min,cur_max
            cur_max=max(nums[i],cur_max*nums[i])
            cur_min=min(nums[i],cur_min*nums[i])
            max_p=max(max_p,cur_max)
        return max_p

File: src/data/courses/dsa/dynamic-programming/state-decision-dp/maximum-product-subarray/code/cpp/maximum_product_subarray.cpp
#include <vector>
#include <algorithm>
#include <climits>
using namespace std;
// ============================================================
// Solution 1: Brute Force — All subarrays — O(N²) Time
// ============================================================
class Solution1 {
public:
    int maxProduct(vector<int>& nums) {
        int res = INT_MIN, n = nums.size();
        for (int i = 0; i < n; i++) {
            int prod = 1;
            for (int j = i; j < n; j++) { prod *= nums[j]; res = max(res, prod); }
        }
        return res;
    }
};

// ============================================================
// Solution 2: Track curMax and curMin — O(N) Time, O(1) Space
// ============================================================
class Solution2 {
public:
    int maxProduct(vector<int>& nums) {
        int res = nums[0], curMax = nums[0], curMin = nums[0];
        for (int i = 1; i < (int)nums.size(); i++) {
            if (nums[i] < 0) swap(curMax, curMin);
            curMax = max(nums[i], curMax * nums[i]);
            curMin = min(nums[i], curMin * nums[i]);
            res = max(res, curMax);
        }
        return res;
    }
};
